lib/Dialect/Comb/CombFolds.cpp

```diff
@@ -239,16 +239,101 @@ OpFoldResult ShlOp::fold(ArrayRef<Attribute> operands) {
       operands, [](APInt a, APInt b) { return a.shl(b); });
 }
 
+LogicalResult ShlOp::canonicalize(ShlOp op, PatternRewriter &rewriter) {
+  // ShlOp(x, cst) -> Concat(Extract(x), zeros)
+  APInt value;
+  if (!matchPattern(op.rhs(), m_RConstant(value)))
+    return failure();
+
+  unsigned srcWidth = op.lhs().getType().cast<IntegerType>().getWidth();
+  unsigned shift = value.getZExtValue();
+
+  if (srcWidth <= shift) {
+    rewriter.replaceOpWithNewOp<hw::ConstantOp>(op,
+                                                APInt::getNullValue(srcWidth));
+    return success();
+  }
+
+  auto zeros =
+      rewriter.create<hw::ConstantOp>(op.getLoc(), APInt::getNullValue(shift));
+
+  auto extractType = rewriter.getIntegerType(srcWidth - shift);
+  auto extract =
+      rewriter.create<ExtractOp>(op.getLoc(), extractType, op.lhs(), shift);
+
+  SmallVector<Value> operand = {extract, zeros};
+  rewriter.replaceOpWithNewOp<ConcatOp>(op, operand);
```

Comment:
You might try:

`rewriter.replaceOpWithNewOp<ConcatOp>(op, ArrayRef<Value>{extract, zeros});`

If that doesn't work, you can use a C array instead of smallvector:

`Value operands[] = {extract, zeros};`

Reply:
Thanks it works